### ISSUE TITLE
FTP.ftp_RETR() should pass FTPCmdError errors back to the client

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4913]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4913|
|Type|enhancement|
|Created|2011-02-25 14:29:17Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-retr-errors-4913|

This is like [#4909](https://github.com/twisted/twisted/issues/4909).  Any error encountered while reading a file and sending it to a client, to respond to a _RETR_ command, gets logged and a `CNX_CLOSED_TXFR_ABORTED` is sent back to the client.  If the failure is an `FTPCmdError` though, it can be presented to the client to give them more information about what went wrong.

Attachments:

* [4913-ftp-stor-err-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/100/1000378ccce258aec46ebaae380f0dd66e8a8e44/4d51f44b34aba1ab90b3398f12b58fd04603d6fc.diff) (2126 bytes) - added by adiroiban on 2012-06-14 00:12:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4913 4913
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_ftp_retr_errors_4913 branches/ftp-retr-errors-4913
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1298644157000000 1298644157000000
changetime__1353247827000000 1353247827000000
version__None None
owner__spiv spiv
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

I have attached a patch based on the implementation from [#4909](https://github.com/twisted/twisted/issues/4909)

Please let me know if it requires changes.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|spiv set owner to spiv|
|-|-|

This looks acceptable to me.  The code for ftp_RETR is alarmingly similar to ftp_STOR, and this leads to annoying similar tests, unsurprisingly… but that's not the fault of this patch, and in fact this patch makes the situation slightly better by making ftp_RETR more similar to ftp_STOR.  So if buildbot is happy with this patch I'll land it for you.

These two methods almost certainly would be better off refactored into a common helper (perhaps involving an actual object with state rather than a bunch of nested functions) that can be unit tested thoroughly (once), but that will be a separate ticket.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|spiv commented|
|-|-|

(In [34782]) Branch to ftp-retr-errors-4913.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spiv set status to `closed`|
|-|-|

(In [34785]) Merge ftp-retr-errors-4913: Pass FTPCmdErrors during FTP.ftp_RETR back to the client.

Author: adiroiban
Reviewer: spiv
Fixes: [#4913](https://github.com/twisted/twisted/issues/4913)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

Thanks for the review and sorry for the late reply.

I agree that there is some code repetition. I plan to fix that as part of [#4180](https://github.com/twisted/twisted/issues/4180)

Cheers,
Adi
